### PR TITLE
Implement CSS alignment specification for absolutely positioned boxes

### LINKED ISSUE #34274: Implement default overflow alignment for abspos
See https://drafts.csswg.org/css-align/#auto-safety-position

We are currently handling the default overflow alignment as unsafe.

Testcase:

```html
<!DOCTYPE html>
<style>
.wrapper { position: relative; width: 50px; height: 50px; border: 3px solid; margin: 75px; }
.abspos { position: absolute; place-self: unsafe end; background: cyan; height: 100px; width: 100px; inset: 25px; }
</style>
<div class="wrapper"><div class="abspos" style="place-self: unsafe end"></div></div>
<div class="wrapper"><div class="abspos" style="place-self: safe end"></div></div>
<div class="wrapper"><div class="abspos" style="place-self: end"></div></div>
```

Expected:

### PR DESCRIPTION
**Pull request by chickenleaf**
**Issue:**
The current implementation doesn't fully handle the default overflow alignment for absolutely positioned elements according to the CSS Alignment Module specification.

**Changes:**
-  Added logic for default overflow alignment in absolutely positioned boxes.
-  Included handling for:
             1.  Safe alignment
             2.  Unsafe alignment
             3.  Stretch and auto alignment
-  Calculates the default overflow rect.

**Specification Implementation:**
Follows the [CSS Alignment Module specification (4.4.1.2)](https://drafts.csswg.org/css-align/#auto-safety-position):
- Computes the default overflow rect.
- Handles overflow:
         - Aligns as specified if within the default overflow rect.
         - Shifts alignment to stay within the rect.
         - Start-aligns if the overflow exceeds the default rect.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #34274

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ____

### CODE REVIEW

**Comment by chickenleaf** [comment]
A lot of tests are failing...

File: components/layout_2020/positioned.rs [hunk 1]
```diff
@@ -940,7 +945,33 @@ impl<'a> AbsoluteAxisSolver<'a> {
...
+                                    .min(Au::zero()) -
+                                margin_box_axis.length,
+                        )
+                    } else {
+                        Some(alignment_container.origin.min(Au::zero()))
+                    }
+                }
+            },
+            AlignFlags::START => Some(alignment_container.origin),
+            AlignFlags::STRETCH | AlignFlags::AUTO => Some(
```
Comment (Loirooriol) [inline components/layout_2020/positioned.rs:974]: Why are you changing `stretch` and `auto`?

**Review by Loirooriol — CHANGES_REQUESTED** [review]

File: components/layout_2020/positioned.rs (hunk 1 above)
Comment (chickenleaf) [reply components/layout_2020/positioned.rs:974]: oh no that is incorrect, i think i interpreted the instructions wrong, `stretch` and `auto` should be handled by the fallback case. i shall remove it

**Comment by chickenleaf** [comment]
a lot of the fails have been fixed, im working on the rest (around 16 more to fix) and also 3 timeouts are showing up, hopefully i send in another commit tomorrow. the unnecessary explicit `stretch` and `auto` has also been removed.

**Comment by chickenleaf** [comment]
so current number of test-fails have been brought down to:
1. `FAIL [expected PASS] ./mach test-wpt /css/CSS2/visuren/right-offset-position-fixed-001.xht`
2. `FAIL [expected PASS] ./mach test-wpt /css/CSS2/normal-flow/block-non-replaced-width-004.xht`
3.  under `./mach test-wpt /css/CSS2/positioning/`

```
▶ OK /css/CSS2/positioning/relpos-percentage-left-in-scrollable-2.html
  │   ▶ FAIL [expected PASS] Left percentage resolved correctly for overflow contribution
  │   │   → assert_equals: expected 0 but got 123456
  │   │
  │   │ @http://web-platform.test:8000/css/CSS2/positioning/relpos-percentage-left-in-scrollable-2.html:21:20
  │   │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2638:25
  │   │ test@http://web-platform.test:8000/resources/testharness.js:633:30
  └   └ @http://web-platform.test:8000/css/CSS2/positioning/relpos-percentage-left-in-scrollable-2.html:20:7

  ▶ FAIL [expected PASS] /css/CSS2/positioning/absolute-non-replaced-width-021.xht
  │   → /css/CSS2/positioning/absolute-non-replaced-width-021.xht ['ca55776a5aa17d37de2e656bd6b957e8924571c1']
  └   → /css/CSS2/positioning/absolute-non-replaced-width-021-ref.xht ['f208affdd0a4f8583c056a78617540201b3c9308']

  ▶ FAIL [expected PASS] /css/CSS2/positioning/absolute-non-replaced-width-024.xht
  │   → /css/CSS2/positioning/absolute-non-replaced-width-024.xht ['ca55776a5aa17d37de2e656bd6b957e8924571c1']
  └   → /css/CSS2/positioning/absolute-non-replaced-width-021-ref.xht ['f208affdd0a4f8583c056a78617540201b3c9308']

  ▶ FAIL [expected PASS] /css/CSS2/positioning/absolute-non-replaced-width-022.xht
  │   → /css/CSS2/positioning/absolute-non-replaced-width-022.xht ['ca55776a5aa17d37de2e656bd6b957e8924571c1']
  └   → /css/CSS2/positioning/absolute-non-replaced-width-021-ref.xht ['f208affdd0a4f8583c056a78617540201b3c9308']

  ▶ FAIL [expected PASS] /css/CSS2/positioning/absolute-non-replaced-width-023.xht
  │   → /css/CSS2/positioning/absolute-non-replaced-width-023.xht ['ca55776a5aa17d37de2e656bd6b957e8924571c1']
  └   → /css/CSS2/positioning/absolute-non-replaced-width-021-ref.xht ['f208affdd0a4f8583c056a78617540201b3c9308']

  ▶ FAIL [expected PASS] /css/CSS2/positioning/abspos-block-level-001.html
  │   → /css/CSS2/positioning/abspos-block-level-001.html ['b4b2feba6434fb0b23d031ec4781e90b080bb0ed']
  └   → /css/CSS2/positioning/abspos-block-level-001-ref.html ['fd228b4aa47ee4cc5577de81c20cf1296ce46cdd']

  ▶ FAIL [expected PASS] /css/CSS2/positioning/auto-position-rtl-child-viewport-scrollbar.html
  │   → /css/CSS2/positioning/auto-position-rtl-child-viewport-scrollbar.html ['9bcb1e53fa857453d72aeb5be938bb89a0cb0fa3']
  └   → /css/CSS2/positioning/auto-position-rtl-child-viewport-scrollbar-ref.html ['d93c9a888fda9e9f9b644c6d90e54cc0532bb072']

  ▶ FAIL [expected PASS] /css/CSS2/positioning/position-relative-035.xht
  │   → /css/CSS2/positioning/position-relative-035.xht ['4f7d536c99c849e0ce9aebdcf01755cb6d129fb1']
  └   → /css/CSS2/positioning/position-relative-035-ref.xht ['e52b8479825c4cc47434ed4b7fa01fd1cf4bd8d9']

  ▶ FAIL [expected PASS] /css/CSS2/positioning/positioning-float-001.xht
  │   → /css/CSS2/positioning/positioning-float-001.xht ['d2ae8910587f432dde42bc69cf5c69b8ef47429e']
  └   → /css/CSS2/positioning/positioning-float-001-ref.xht ['75bf03ca8353a9ea1c47d8728a6898bee074d49a']

Ran 526 tests finished in 380.6 seconds.
  • 8 ran as expected.
  • 8 tests failed unexpectedly
  • 1 tests had unexpected subtest results
```
and TIMEOUTS under `./mach test-wpt /css/css-flexbox/abspos/` :
```
  ▶ TIMEOUT [expected OK] /css/css-flexbox/abspos/position-absolute-013.html

  ▶ TIMEOUT [expected OK] /css/css-flexbox/abspos/position-absolute-012.html

  ▶ TIMEOUT [expected OK] /css/css-flexbox/abspos/position-absolute-001.html

Ran 65 tests finished in 48.3 seconds.
  • 8 ran as expected.
  • 3 tests timed out unexpectedly
```
i am kind of unsure how to proceed next, what other edge-cases need to be considered?

**Comment by chickenleaf** [comment]
> Can you push the update?

Done

File: components/layout_2020/positioned.rs [hunk 2]
```diff
@@ -940,9 +944,24 @@ impl<'a> AbsoluteAxisSolver<'a> {
...
+                } else if margin_box_axis.length <= default_overflow_rect {
+                    Some(
+                        alignment_container.origin.max(Au::zero()) + alignment_container.length -
+                            margin_box_axis.length,
+                    )
+                } else {
+                    Some(Au::zero())
+                }
+            },
+            AlignFlags::START => Some(alignment_container.origin),
```
Comment (Loirooriol) [inline components/layout_2020/positioned.rs:964]: Why are you changing `AlignFlags::START`? It breaks static position. This should be handled by `_ => None`.

File: components/layout_2020/positioned.rs [hunk 3]
```diff
@@ -928,9 +928,13 @@ impl<'a> AbsoluteAxisSolver<'a> {
...
         };
 
+        let default_overflow_rect = (alignment_container.origin.max(Au::zero())) +
+            (alignment_container.length.max(Au::zero()));
+
         let mut value_after_safety = self.alignment.value();
         if self.alignment.flags() == AlignFlags::SAFE &&
-            margin_box_axis.length > alignment_container.length
+            (margin_box_axis.length > alignment_container.length ||
+                alignment_container.origin < Au::zero())
```
Comment (Loirooriol) [inline components/layout_2020/positioned.rs:937]: Don't force start alignment just because the start inset is negative.

```html
<!DOCTYPE html>
<div style="margin: 40px; border: solid; position: relative; width: 100px; height: 100px">
  <div style="position: absolute; z-index: -1; inset: -20px; background: cyan; place-self: safe end; width: 50px; height: 50px"></div>
</div>
```

**Review by Loirooriol — CHANGES_REQUESTED** [review]

**Review by Loirooriol — CHANGES_REQUESTED** [review]

File: components/layout_2020/positioned.rs [hunk 4]
```diff
@@ -928,9 +928,13 @@ impl<'a> AbsoluteAxisSolver<'a> {
             _ => return None,
         };
 
+        let default_overflow_rect = (alignment_container.origin.max(Au::zero())) +
+            (alignment_container.length.max(Au::zero()));
```
[the remark this answers is not in the document]
Comment (Loirooriol) [reply components/layout_2020/positioned.rs:932]: Disregard the last part, for static position it's not necessarily `auto`.
I think then it would be

```rs
self.containing_size.max(static_position_rect_axis.origin + static_position_rect_axis.length)
    - static_position_rect_axis.origin.min(Au::zero())
```

A helper method for `RectAxis` that returns the end position could be useful.

**Comment by Loirooriol** [comment]
Done in #35208
